### PR TITLE
feat(annotation-thread): create base for AnnotationThread component

### PR DESCRIPTION
**Pull request by wpiesiak**
Create an `AnnotationThread` component to be reusable outside Sidebar.
Implement base functionality to render `AnnotationActivity`.
Create withCollaborators HOC to make collaborators related functions reusable.
Create useAnnotationAPI hook for base api annotation requests.

In next steps will be implemented: 
- functionality to create annotation if there is no id passed in props

### CODE REVIEW

File: src/elements/annotation-thread/AnnotationThread.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/**
+ * @flow
```
Comment (ConradJChan) [inline src/elements/annotation-thread/AnnotationThread.js:2]: We're trying to migrate box-ui-elements to typescript so new development should be written in typescript, if possible

File: src/elements/annotation-thread/AnnotationThread.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+ * @file Annotation Thread Container
+ * @author Box
+ */
+
+import React from 'react';
+import flow from 'lodash/flow';
```
Comment (ConradJChan) [inline src/elements/annotation-thread/AnnotationThread.js:8]: Do we need `flow` currently?

**Review by ConradJChan — COMMENTED** [review]
Should this be a top level element like `ContentExplorer`, `ContentPreview`, etc...? Seems like putting it under `src/features/annotations` might be more appropriate?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    const getMentionWithQuery = debounce(
+        (searchStr: string) => getCollaborators(getMentionContactsSuccessCallback, errorCallback, searchStr),
+        DEFAULT_COLLAB_DEBOUNCE,
+    );
+
+    const getAvatarUrl = async (userId: string): Promise<?string> => {
+        return api.getUsersAPI(false).getAvatarUrlWithAccessToken(userId, fileId);
+    };
+
+    const getUserProfileUrl = (userID: string) => Promise.resolve(`/profile/${userID}`);
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThreadContent.js:164]: `ActivitySidebar` seems to have similar functionality, but calls
```
getAvatarUrl = async (userId: string): Promise<?string> => {
        const { file, api } = this.props;

        return api.getUsersAPI(false).getAvatarUrlWithAccessToken(userId, file.id);
    };
```
Would that work here?

File: src/features/annotation-thread/AnnotationThread.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        token,
+    });
+
+    if (!annotationId) {
+        return null;
+    }
+
+    return (
+        <Internationalize language={language} messages={messages}>
+            <APIContext.Provider value={(api: any)}>
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThread.js:67]: Is the `APIContext` needed here if we're passing the `api` down one level?

File: src/features/annotation-thread/AnnotationThread.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+                <div className={`be bat ${className}`}>
```
[the remark this answers is not in the document]
Comment (tjuanitas) [reply src/features/annotation-thread/AnnotationThread.js:68]: We could probably remove `be` since it's no longer under elements. Should `bat` be `AnnotationThread` instead?

File: src/features/annotation-thread/AnnotationThread.scss [hunk 6]
```diff
@@ -0,0 +1,10 @@
+@import 'src/elements/common/variables';
+
+$annotationThreadWidth: 340px;
+$annotationThreadHeight: 600px;
```
Comment (tjuanitas) [inline src/features/annotation-thread/AnnotationThread.scss:4]: These could be SUIT styled: `$AnnotationThread-width`

https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md#variables

File: src/features/annotation-thread/AnnotationThread.scss [hunk 7]
```diff
@@ -0,0 +1,10 @@
+@import 'src/elements/common/variables';
```
Comment (tjuanitas) [inline src/features/annotation-thread/AnnotationThread.scss:1]: Since it's no longer under `elements`, we could use the `src/styles/variables` file.

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+        );
+    };
+
+    React.useEffect(() => {
+        fetchAnnotation();
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [annotationId]);
+
+    const annotationSuccessCallback = () => {
+        fetchAnnotation();
```
[the remark this answers is not in the document]
Comment (wpiesiak) [reply src/features/annotation-thread/AnnotationThreadContent.js:94]: It will, but only annotation without replies. Optionally we could add later on possibility to return annotation with all its replies during update.

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    };
+
+    const handleAnnotationDelete = ({ id, permissions }: { id: string, permissions: AnnotationPermission }) => {
+        api.getAnnotationsAPI(false).deleteAnnotation(fileId, id, permissions, noop, annotationErrorCallback);
```
[the remark this answers is not in the document]
Comment (wpiesiak) [reply src/features/annotation-thread/AnnotationThreadContent.js]: added handling `isPending` attribute on Annotation which will be handled by `AnnotationActivity`

File: src/features/annotation-thread/AnnotationThreadContent.js (hunk 3 above)
Comment (wpiesiak) [reply src/features/annotation-thread/AnnotationThreadContent.js]: this one is for getting Avatar. `getUserProfileUrl` should get user profile and actually the function should be exposed as external prop in `AnnotationThread`, I fixed it 👍

File: src/features/annotation-thread/AnnotationThread.js (hunk 4 above)
Comment (wpiesiak) [reply src/features/annotation-thread/AnnotationThread.js:71]: I wanted to keep consistency with Sidebar if some common components in lower levels would need api context currently or in the future.

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 10]
```diff
@@ -48,13 +46,14 @@ import type {
...
-import type { ElementsErrorCallback, ErrorContextProps, ElementsXhrError } from '../../common/types/api';
+import type { ErrorContextProps, ElementsXhrError } from '../../common/types/api';
 import type { WithLoggerProps } from '../../common/types/logging';
 import type { SelectorItems, User, UserMini, GroupMini, BoxItem } from '../../common/types/core';
 import type { GetProfileUrlCallback } from '../common/flowTypes';
-import type { Translations, Collaborators, Errors } from './flowTypes';
+import type { Translations, Errors } from './flowTypes';
 import type { FeatureConfig } from '../common/feature-checking';
 import './ActivitySidebar.scss';
+import withCollaborators from '../common/collaborators/withCollaborators';
```
Comment (ConradJChan) [inline src/elements/content-sidebar/ActivitySidebar.js:56]: Can we move this import above to be sorted appropriately?

File: src/elements/content-sidebar/ActivitySidebar.js [hunk 11]
```diff
@@ -638,19 +637,18 @@ class ActivitySidebar extends React.PureComponent<Props, State> {
...
      */
-    getApproverWithQuery = debounce(
-        (searchStr: string) =>
-            this.getCollaborators(this.getApproverContactsSuccessCallback, this.errorCallback, searchStr, {
-                includeGroups: true,
-            }),
-        DEFAULT_COLLAB_DEBOUNCE,
-    );
+    getApprover = (searchStr: string) => {
+        const { file, getCollaboratorsWithQuery } = this.props;
```
Comment (ConradJChan) [inline src/elements/content-sidebar/ActivitySidebar.js:647]: Do we need to add `getCollaboratorsWithQuery` to the expected `Props` type?

File: src/elements/content-sidebar/__tests__/ActivitySidebar.test.js [hunk 12]
```diff
@@ -676,33 +676,25 @@ describe('elements/content-sidebar/ActivitySidebar', () => {
         });
     });
 
-    describe('getApproverWithQuery()', () => {
+    describe('getApprover()', () => {
         let instance;
         let wrapper;
-        let getCollaboratorsSpy;
+        const getCollaboratorsWithQuery = jest.fn();
```
Comment (ConradJChan) [inline src/elements/content-sidebar/__tests__/ActivitySidebar.test.js:682]: Can we move the `const` declarations above the `let` declarations?

File: src/features/annotation-thread/AnnotationThread.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+import type APICache from '../../utils/Cache';
+import type { BoxItemPermission, StringMap, Token } from '../../common/types/core';
+import type { ErrorContextProps } from '../../common/types/api';
+
+import './AnnotationThread.scss';
+
+type Props = {
+    annotationId?: string,
+    apiHost?: string,
```
[the remark this answers is not in the document]
Comment (ConradJChan) [reply src/features/annotation-thread/AnnotationThread.js:25]: I think there's two ways to interpret "top level component":
(1) official BUIE element (i.e. ContentPreview, ContentExplorer, etc...)
(2) the component imported from BUIE into a client application

For (1) we need to add these props of `apiHost` for 3rd party developers to have the flexibility to use, etc... but for (2) especially in the `features` directory it can be anything you want for your 1st party application

If you are using `apiHost`, `requestInterceptor`, etc... then ignore this, but if you're not, then I think we can exclude it

File: src/constants.js [hunk 14]
```diff
@@ -310,6 +310,7 @@ export const ORIGIN_PREVIEW: 'preview' = 'preview';
 export const ORIGIN_PREVIEW_SIDEBAR: 'preview_sidebar' = 'preview_sidebar';
 export const ORIGIN_CONTENT_EXPLORER: 'content_explorer' = 'content_explorer';
 export const ORIGIN_OPEN_WITH: 'open_with' = 'open_with';
+export const ORIGIN_ANNOTATION_THREAD: 'annotation_thread' = 'annotation_thread';
```
Comment (jstoffan) [inline src/constants.js:313]: Let's revert this change. These are intended for use by top-level Elements like ContentSidebar.

File: src/elements/common/collaborators/__tests__/withCollaborators.test.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        getFileCollaboratorsAPI: () => fileCollaboratorsAPI,
+    };
+    const defaultProps = { api };
+    const getWrapper = props => shallow(<WrappedComponent {...defaultProps} {...props} />);
+    let fileId;
+    let wrapper;
+    let instance;
+    let successCb;
+    let errorCb;
+    beforeEach(() => {
```
Comment (jstoffan) [inline src/elements/common/collaborators/__tests__/withCollaborators.test.js:23]: Can we add some whitespace between logical blocks, similar to our other test files?

File: src/features/annotation-thread/AnnotationThread.scss [hunk 16]
```diff
@@ -0,0 +1,10 @@
+@import 'src/styles/variables';
+
+$annotationThread-width: 340px;
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThread.scss:3]: I believe we're following SUIT style naming for our sass variables, should this be `$AnnotationThread-width`?

File: src/features/annotation-thread/AnnotationThread.js [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    });
+
+    if (!annotationId) {
+        return null;
+    }
+
+    return (
+        <Internationalize language={language} messages={messages}>
+            <APIContext.Provider value={(api: any)}>
+                <div className={classNames('bat', className)}>
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThread.js:72]: Could this have a class `AnnotationThread` instead of `bat`?

File: src/elements/common/collaborators/withCollaborators.js [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+
+type Props = {
+    api: API,
+};
+
+export default function withCollaborators(WrappedComponent: React.ComponentType<Props>): React.ComponentType<Props> {
+    class WithMentions extends React.Component<Props> {
+        static displayName: ?string;
+
+        props: Props;
```
Comment (jstoffan) [inline src/elements/common/collaborators/withCollaborators.js:18]: I think `React.Component` should take care of this for us.

File: src/elements/common/collaborators/withCollaborators.js [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+import API from '../../../api/APIFactory';
+import type { Collaborators } from '../../content-sidebar/flowTypes';
+import type { ElementsErrorCallback } from '../../../common/types/api';
+
+type Props = {
+    api: API,
+};
+
+export default function withCollaborators(WrappedComponent: React.ComponentType<Props>): React.ComponentType<Props> {
+    class WithMentions extends React.Component<Props> {
```
Comment (jstoffan) [inline src/elements/common/collaborators/withCollaborators.js:15]: Is this class name correct?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+const AnnotationThreadContent = ({
+    annotationId,
+    api,
+    fileId,
+    getCollaboratorsWithQuery,
+    getUserProfileUrl,
+    filePermissions,
+    onError,
+}: Props) => {
+    const [mentionSelectorContacts, setMentionSelectorContacts] = React.useState();
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThreadContent.js:38]: Should this have a default value?

File: src/features/annotation-thread/AnnotationThread.js [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+ * @flow
+ * @file Annotation Thread Container
+ * @author Box
+ */
+
+import React from 'react';
+import classNames from 'classnames';
+import AnnotationThreadContent from './AnnotationThreadContent';
+import API from '../../api/APIFactory';
+import APIContext from '../../elements/common/api-context';
```
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThread.js:11]: It's not ideal for a feature-level component to import modules from the `elements` directory. It doesn't seem like APIContext or Internationalize add a ton of value here. Can we pass the API instance down directly as a prop and use `IntlProvider`, instead?

File: src/features/annotation-thread/AnnotationThread.js [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+import Internationalize from '../../elements/common/Internationalize';
+import { DEFAULT_HOSTNAME_API, ORIGIN_ANNOTATION_THREAD } from '../../constants';
+import { withErrorBoundary } from '../../elements/common/error-boundary';
```
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThread.js:14]: I would suggest adding an error boundary in the consuming application rather than at the component level.

File: src/features/annotation-thread/AnnotationThread.js [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+type Props = {
+    annotationId?: string,
+    apiHost?: string,
+    cache?: APICache,
+    className: string,
+    clientName: string,
+    fileId: string,
+    filePermissions: BoxItemPermission,
+    getUserProfileUrl?: GetProfileUrlCallback,
+    isDefaultOpen?: boolean,
```
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThread.js:32]: It looks like some of these props are unused currently. Can we remove anything that's not necessary?

File: src/features/annotation-thread/AnnotationThread.scss (hunk 16 above)
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThread.scss:3]: Variables at this level are global. Since these values are only used once and only in this file, do they need to be pulled out as variables?

File: src/features/annotation-thread/AnnotationThread.js (hunk 17 above)
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThread.js:72]: We typically reserve this naming convention for top-level Elements like ContentSidebar. Can we use SUIT naming here?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 24]
```diff
@@ -0,0 +1,180 @@
+// @flow
+import React from 'react';
+import flow from 'lodash/flow';
+import ActivityThread from '../../elements/content-sidebar/activity-feed/activity-feed/ActivityThread';
+import AnnotationActivity from '../../elements/content-sidebar/activity-feed/annotations';
+import API from '../../api/APIFactory';
+import commonMessages from '../../elements/common/messages';
+import LoadingIndicator from '../../components/loading-indicator/LoadingIndicator';
+import { withAPIContext } from '../../elements/common/api-context';
+import withErrorHandling from '../../elements/content-sidebar/withErrorHandling';
```
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThreadContent.js:10]: We're introducing a lot of dependencies on modules from the `elements` directory, where typically we see the inverse, with the top-level Elements using some components from the `features` directory.

File: src/features/annotation-thread/AnnotationThreadContent.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ConradJChan) [reply src/features/annotation-thread/AnnotationThreadContent.js:94]: Seems like it's unnecessary to fetch all the replies if all we're updating is the annotation status or message?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+        setAnnotationError(undefined);
+        setIsLoading(false);
+    };
+
+    const errorCallback = React.useCallback(
+        (error: ElementsXhrError, code: string, contextInfo: Object = {}): void => {
+            /* eslint-disable no-console */
+            console.error(error);
+            /* eslint-enable no-console */
+            onError && onError(error, code, contextInfo);
```
Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThreadContent.js:54]: Can we use a standard `if` block guard clause here?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+        fetchAnnotation();
+    }, [annotationId, fetchAnnotation]);
+
+    const annotationSuccessCallback = () => {
+        setAnnotation({ ...annotation, isPendind: false });
+        fetchAnnotation();
+    };
+
+    const annotationDeleteSuccessCallback = () => {
+        setAnnotation({ ...annotation, isPendind: false });
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThreadContent.js:98]: spelling `isPendind` -> `isPending`?

Comment (jstoffan) [inline src/features/annotation-thread/AnnotationThreadContent.js:98]: Typo: `isPendind` -> `isPending`

File: src/features/annotation-thread/__tests__/AnnotationThreadContent.test.js [hunk 27]
```diff
@@ -0,0 +1,64 @@
+// @flow
+
+import { mount } from 'enzyme';
+import React from 'react';
+import { AnnotationThreadContentComponent } from '../AnnotationThreadContent';
+import { annotation } from '../../../__mocks__/annotations';
+import ActivityThread from '../../../elements/content-sidebar/activity-feed/activity-feed/ActivityThread';
+import AnnotationActivity from '../../../elements/content-sidebar/activity-feed/annotations';
+
+describe('elements/annotation-thread/AnnotationThreadContent', () => {
```
Comment (jstoffan) [inline src/features/annotation-thread/__tests__/AnnotationThreadContent.test.js:10]: Should this path be updated or removed? This test suite also seems quite light given the amount of logic in AnnotationThreadContent.

File: src/elements/common/collaborators/withCollaborators.js [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+         *
+         * @param {string} fileId
+         * @param {Function} successCallback - the success callback
+         * @param {Function} errorCallback - the error callback
+         * @param {string} searchStr - the search string
+         * @param {Object} [options]
+         * @param {boolean} [options.includeGroups] - return groups as well as users
+         * @return {void}
+         */
+        getCollaborators = (
```
Comment (jstoffan) [inline src/elements/common/collaborators/withCollaborators.js:31]: If this is a commonly-needed behavior, does it need to live in a separate HOC or can we augment the `FileCollaborators` API module, instead?

File: src/elements/common/collaborators/__tests__/withCollaborators.test.js [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+        fileId = '1';
+        successCb = jest.fn();
+        errorCb = jest.fn();
+        wrapper = getWrapper({
+            api,
+        });
+        instance = wrapper.instance();
+    });
+    describe('getCollaborators', () => {
+        test('should short circuit if there is no search string', () => {
```
Comment (jstoffan) [inline src/elements/common/collaborators/__tests__/withCollaborators.test.js:33]: Should this be a `test.each`?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 30]
```diff
@@ -0,0 +1,180 @@
...
+    };
+
+    const handleAnnotationDelete = ({ id, permissions }: { id: string, permissions: AnnotationPermission }) => {
+        setAnnotation({ ...annotation, isPendind: true });
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThreadContent.js:102]: Seems like it's possible to have both the `Sidebar` and `AnnotationThread` components open at the same time. What is the expected experience when any of these delete/edit/resolve annotations actions are made? Will the sidebar also reflect a pending state for that given annotation? The Feed api manages it's own pending state which is separate from the state being managed here

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 31]
```diff
@@ -0,0 +1,180 @@
...
+
+    const getMentions = (searchStr: string) =>
+        getCollaboratorsWithQuery(fileId, getMentionContactsSuccessCallback, errorCallback, searchStr);
+
+    const getAvatarUrl = async (userId: string): Promise<?string> => {
+        return api.getUsersAPI(false).getAvatarUrlWithAccessToken(userId, fileId);
+    };
+
+    if (annotationError) {
+        return null;
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThreadContent.js:155]: This this a placeholder for a future error state?

File: src/features/annotation-thread/AnnotationThreadContent.js [hunk 32]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    return (
+        <ActivityThread hasReplies>
+            {isLoading && <LoadingIndicator />}
+            {!isLoading && !annotationError && annotation && (
```
Comment (ConradJChan) [inline src/features/annotation-thread/AnnotationThreadContent.js:161]: Is this needed if there is a shortcut for `annotationError` above?

File: src/constants.js (hunk 14 above)
Comment (wpiesiak) [reply src/constants.js:313]: I believe there is some misunderstanding - AnnotationThread is meant to be a top level component. The plan for AnnotationThread is to exist independently form Sidebar, wrap AnnotationActivity with all its functionalities and possibility to create new Annotation as well (will be added in next PRs)

File: src/elements/common/collaborators/withCollaborators.js [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+
+import * as React from 'react';
+import debounce from 'lodash/debounce';
+import { DEFAULT_COLLAB_DEBOUNCE } from '../../../constants';
+import API from '../../../api/APIFactory';
+import type { Collaborators } from '../../content-sidebar/flowTypes';
+import type { ElementsErrorCallback } from '../../../common/types/api';
+
+type Props = {
+    api: API,
```
[the remark this answers is not in the document]
Comment (wpiesiak) [reply src/elements/common/collaborators/withCollaborators.js:11]: the `api` prop is passed from `withAPIContext`. There is no need to pass it as an argument. Similar approach exists currently in `withSidebarAnnotation`

File: src/features/annotation-thread/AnnotationThreadContent.js (hunk 31 above)
Comment (wpiesiak) [reply src/features/annotation-thread/AnnotationThreadContent.js:155]: more like for create new annotation state. If there is no annotationId, component should render form and handle create new annotation.

File: src/features/annotation-thread/AnnotationThreadContent.js (hunk 30 above)
Comment (wpiesiak) [reply src/features/annotation-thread/AnnotationThreadContent.js:102]: Good call, I will consider this case while creating event bus for sharing state between Sidebar and AnnotationThread

File: src/constants.js (hunk 14 above)
Comment (jstoffan) [reply src/constants.js:313]: Discussed offline, but to capture here: We can have top-level components/features that aren't top-level UI Elements. It looks like this new component is the former rather than the latter.

File: src/elements/common/collaborators/withCollaborators.js (hunk 28 above)
Comment (wpiesiak) [reply src/elements/common/collaborators/withCollaborators.js]: I think HOC helps with handling additional debounce logic. If you don't mind, I would rather to keep it that way

File: src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js [hunk 34]
```diff
@@ -0,0 +1,87 @@
...
+import type { BoxItemPermission, SelectorItems } from '../../../../common/types/core';
+import type { WithCollaboratorsProps } from '../../../common/collaborators/withCollaborators';
+import useAnnotationAPI from './useAnnotationAPI';
+
+type Props = {
+    annotationId: string,
+    api: API,
+    fileId: string,
+    filePermissions: BoxItemPermission,
+    onError: (error: ElementsXhrError | Error, code: string, contextInfo?: Object, origin?: ElementOrigin) => void,
```
[the remark this answers is not in the document]
Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js]: I have no idea why but ESLint has trouble with `& ErrorContextProps` in this particular file 🤯 
Screenshot 2022-09-22 at 15 13 45

Used `onError: ErrorContextProps.onError `instead

File: src/elements/common/collaborators/withCollaborators.js [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @param {string} fileId
+     * @param {Function} successCallback - the success callback
+     * @param {Function} errorCallback - the error callback
+     * @param {string} searchStr - the search string
+     * @param {Object} [options]
+     * @param {boolean} [options.includeGroups] - return groups as well as users
+     * @return {void}
+     */
+    const getCollaborators = (
```
[the remark this answers is not in the document]
Comment (wpiesiak) [reply src/elements/common/collaborators/withCollaborators.js]: Moved to FileCollaboratorsAPI 👍

File: src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js [hunk 36]
```diff
@@ -0,0 +1,87 @@
...
+    const getAvatarUrl = async (userId: string): Promise<?string> => {
+        return api.getUsersAPI(false).getAvatarUrlWithAccessToken(userId, fileId);
+    };
+
+    if (isError) {
+        return null;
+    }
+
+    return (
+        <ActivityThread hasReplies getAvatarUrl={getAvatarUrl}>
```
[the remark this answers is not in the document]
Comment (wpiesiak) [reply src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js:53]: Why should it?

File: i18n/en-US.properties [hunk 37]
```diff
@@ -1,3 +1,9 @@
+# Error message when an annotation deletion fails
+annotationThread.errorDeleteAnnotation = There was an error deleting this item.
```
Comment (jstoffan) [inline i18n/en-US.properties:2]: Should these be prefixed with `be.`?

**Review by jstoffan — DISMISSED** [review]
@ConradJChan @wpiesiak, since we're modifying existing components, let's make sure to engage with QA to test the Web Annotations flows.

File: src/api/FileCollaborators.js [hunk 38]
```diff
@@ -8,6 +8,7 @@ import MarkerBasedAPI from './MarkerBasedAPI';
 import { DEFAULT_MAX_COLLABORATORS } from '../constants';
 import type { ElementsErrorCallback } from '../common/types/api';
 import type { SelectorItem, SelectorItems, UserMini, GroupMini } from '../common/types/core';
+import type { Collaborators } from '../elements/content-sidebar/flowTypes';
```
Comment (jstoffan) [inline src/api/FileCollaborators.js:11]: We want to try to avoid importing component modules into our API layer. Can we pull up this type to the "common" modules?

File: src/api/FileCollaborators.js [hunk 39]
```diff
@@ -87,6 +88,36 @@ class FileCollaborators extends MarkerBasedAPI {
...
+     * @param {Function} successCallback
+     * @param {Function} errorCallback
+     * @param {string} searchStr - Search string to filter file collaborators by
+     * @param {Object} [options]
+     * @param {boolean} [options.includeGroups] - return groups as well as users
+     * @return {void}
+     */
+    getCollaboratorsWithQuery = (
+        fileId: string,
+        successCallback: Collaborators => void,
```
Comment (jstoffan) [inline src/api/FileCollaborators.js:105]: Should this share the same type as the success callback in `getFileCollaborators`?

File: src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js [hunk 40]
```diff
@@ -0,0 +1,78 @@
...
+        setMentionSelectorContacts(entries);
+    };
+
+    const getMentions = (searchStr: string) =>
+        api
+            .getFileCollaboratorsAPI(false)
+            .getCollaboratorsWithQuery(fileId, getMentionContactsSuccessCallback, onError, searchStr);
+
+    const getAvatarUrl = async (userId: string): Promise<?string> => {
+        return api.getUsersAPI(false).getAvatarUrlWithAccessToken(userId, fileId);
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js:51]: Above, we're using implicit return, but here we're using explicit return. We also declare the return type in some places but not others. Can we make it consistent?

File: src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js [hunk 41]
```diff
@@ -0,0 +1,78 @@
...
+
+    const { annotation, isLoading, error, handleEdit, handleStatusChange, handleDelete } = useAnnotationAPI({
+        api,
+        fileId,
+        annotationId,
+        filePermissions,
+        errorCallback: onError,
+    });
+
+    /**
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/annotation-thread/AnnotationThreadContent.js:34]: Do we need these JSDocs only for this function?

File: src/elements/content-sidebar/activity-feed/annotation-thread/useAnnotationAPI.js [hunk 42]
```diff
@@ -0,0 +1,155 @@
...
+
+    const annotationDeleteSuccessCallback = () => {
+        // will emit delete event
+    };
+
+    const createAnnotationErrorCallback = (code: string) => {
+        let message;
+        switch (code) {
+            case ERROR_CODE_FETCH_ANNOTATION:
+                message = messages.errorFetchAnnotation;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/annotation-thread/useAnnotationAPI.js:72]: Minor, but we could use an object declared outside the component to map errors to messages.

File: src/elements/content-sidebar/activity-feed/annotation-thread/errors.js [hunk 43]
```diff
@@ -0,0 +1,14 @@
...
+} from '../../../../constants';
+import messages from './messages';
+
+const AnnotationErrors = {
+    [ERROR_CODE_FETCH_ANNOTATION]: messages.errorFetchAnnotation,
+    [ERROR_CODE_EDIT_ANNOTATION]: messages.errorEditAnnotation,
+    [ERROR_CODE_DELETE_ANNOTATION]: messages.errorDeleteAnnotation,
+};
+
+export default AnnotationErrors;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/annotation-thread/errors.js:14]: Can we use camelCase (`annotationErrors`) since this is a regular variable rather than a component?

File: src/elements/content-sidebar/activity-feed/annotation-thread/messages.js [hunk 44]
```diff
@@ -0,0 +1,21 @@
+import { defineMessages } from 'react-intl';
+
+const messages = defineMessages({
+    errorFetchAnnotation: {
+        id: 'be.annotattionThread.errorFetchAnnotation',
+        description: 'Error message when an annotation fetch fails',
+        defaultMessage: 'Cannot fetch annotation.',
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/annotation-thread/messages.js:7]: Minor, but our grammar is a bit inconsistent between errors. Perhaps `The annotation could not be fetched`?

File: src/elements/content-sidebar/activity-feed/annotation-thread/useAnnotationAPI.js [hunk 45]
```diff
@@ -0,0 +1,136 @@
...
+    };
+
+    const annotationDeleteSuccessCallback = () => {
+        // will emit delete event
+    };
+
+    const createAnnotationErrorCallback = (code: string) => {
+        setError({
+            title: commonMessages.errorOccured,
+            message: AnnotationErrors[code],
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/annotation-thread/useAnnotationAPI.js:66]: Should we have a default message if the code isn't present in our map?

**Review by jstoffan — APPROVED** [review]
